### DISCUSSION TITLE
Is there a way to define singleton alt class case

### DISCUSSION THREAD

**Discussion by dadhi** [qwertie/ecsharp#51] [Q&A]
I have a Messages and probably many more abstractions which a best represented as ADT [here](https://github.com/dadhi/Tea).

     public abstract alt class Msg : IMsg<Counter>
        {
            alt Increment(); // here I want a single instance, and better of base type Msg
            alt Decrement();
        }

Btw, kudos for such an interesting project!

**Comment by qwertie** [comment]
Wow, I'm not sure how I didn't notice this message for two days. Sorry.

`alt class` doesn't support singletons or even non-public constructors, but here's what else you can do. I take it you want C# code that looks like this?

~~~csharp
	public abstract class Msg : IMsg<Counter> {
		public Msg() { }
	}
	class Increment : Msg {
		private static Increment _value = new Increment();
		public static Increment Value { get { return _value; } }
		private Increment() { }
		...
	}
	class Decrement : Msg {
		private static Decrement _value = new Decrement();
		public static Decrement Value { get { return _value; } }
		public Decrement() { }
		...
	}
~~~

I typically write repetitive code using the "define" macro:

~~~csharp
public abstract class Msg : IMsg<Counter> {
	public Msg() { }
}

define MakeSingletonClass($ClassName, $BaseClass, {$(..members);}) {
	public class $ClassName : $BaseClass {
		private static $ClassName _value = new $ClassName();
		public static $ClassName Value { get { return _value; } }
		private this() { } // constructor
		$members;
	}
}
MakeSingletonClass(Increment, Msg)
{
    public void MethodOfIMsg() => Console.WriteLine("Incrementor!");
}
MakeSingletonClass(Decrement, Msg)
{
    public void MethodOfIMsg() => Console.WriteLine("Decrementor!");
}
~~~

A couple of notes:

- In EC#, `f(a) {b;}` is a syntactic alias for `f(a, {b;})`, so `MakeSingletonClass(Decrement, Msg) {...}` calls the `MakeSingletonClass($ClassName, $BaseClass, {$(..members);})` macro.
- The semicolons in `{$(..members);}` and `$members;` are important and required.

**Comment by dadhi** [comment]
Thanks for the answer, it helps.

Macros are powerful, and easy to forget about it in C# world :)

Btw, I more more see an examples of moving last argument out of a list to the make a kind of scope.. Kotlin recently. Good for DSLs.
